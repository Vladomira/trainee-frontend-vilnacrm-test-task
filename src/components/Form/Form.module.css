.formContainer {
  padding: 2.25rem 2.563rem 2.25rem;
  border-radius: 2rem 2rem 0px 0px;
  background-color: white;
  box-shadow: rgba(59, 68, 80, 0.05) 1px 1px 41px 0px;
}

.centredFlexContainer {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}

.formComponentWrapper {
  position: relative;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  margin: 50px 0px;
}
.backgroundImg {
  position: absolute;
  width: 100dvw;
  max-width: 24.5rem;
  height: 30.5rem;
}

.saveButton {
  width: 100%;
  padding: 1.25rem 2rem;
  margin-top: 15px;

  font-size: 1.125rem;
  line-height: 1.125;
  font-weight: 600;

  border-radius: 3.563rem;
  box-shadow:
    rgba(0, 0, 0, 0.2) 0px 3px 1px -2px,
    rgba(0, 0, 0, 0.14) 0px 2px 2px 0px,
    rgba(0, 0, 0, 0.12) 0px 1px 5px 0px;
}
