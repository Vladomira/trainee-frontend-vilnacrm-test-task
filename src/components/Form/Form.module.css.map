{"version":3,"sources":["Form.module.scss","Form.module.css"],"names":[],"mappings":"AAeA;EAJI,aAAA;EAAA,uBAAA;EAAA,mBAAA;EAMF,WAAA;EACA,YAAA;EACA,kBAAA;ACZF;;ADeA;EAXI,aAAA;EAAA,uBAAA;EAAA,mBAAA;EAaF,UAAA;EACA,sBAAA;EACA,cAAA;EACA,mBAAA;EACA,uBAAA;EACA,iDA7BW;ACmBb;ADYE;EATF;IAUI,wBAAA;IACA,gBAAA;IACA,eAAA;ECTF;AACF;ADWE;EAfF;IAgBI,gBAAA;IACA,aAAA;ECRF;AACF;ADUE;EApBF;IAqBI,gBAAA;IACA,eAAA;ECPF;AACF;;ADSA;EACE,WAAA;ACNF;ADQE;EAHF;IAII,gBAAA;ECLF;AACF;;ADQA;EACE,gBAAA;EACA,gBAAA;ACLF;;ADQA;EACE,cAAA;EACA,kBAAA;EACA,gBAAA;EACA,WAAA;EACA,gBAAA;EACA,YAAA;EACA,kBAAA;EAEA,mBApES;EAqET,kBApEW;EAqEX,gBAAA;EAEA,iDAzEW;EA0EX,uBAAA;EACA,YAAA;EACA,eAAA;EAEA,+DAAA;ACRF;ADUE;EAEE,mCAAA;ACTJ;;ADaA;EACE,mBAtFS;EAuFT,kBAtFW;EAuFX,mBAAA;ACVF","file":"Form.module.css"}