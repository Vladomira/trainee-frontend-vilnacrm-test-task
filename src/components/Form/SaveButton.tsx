import Button from '@mui/material/Button';
import React, { FormEvent } from 'react';

import styles from './Form.module.css';

type SaveButtonProps = {
  isDisabled: boolean;
  onHandleSubmit: (e: FormEvent<Element>) => Promise<void>;
};

export default function SaveButton({ isDisabled, onHandleSubmit }: SaveButtonProps) {
  return (
    <Button
      variant="contained"
      disabled={isDisabled}
      size="medium"
      type="submit"
      onClick={onHandleSubmit}
      data-testid="save-button"
      className={styles.saveButton}
      sx={{
        backgroundColor: isDisabled ? 'rgb(168, 207, 230)' : 'rgb(30, 174, 255)',
      }}
    >
      Save
    </Button>
  );
}
