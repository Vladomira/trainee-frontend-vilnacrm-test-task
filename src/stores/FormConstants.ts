export const FORM_ERROR_MESSAGES = {
  INVALID_EMAIL: 'Invalid email format',
  EMAIL_REQUIRED: 'Please type email',
  INVALID_PHONE: 'Wrong phone format',
  NAME_REQUIRED: 'Please type your name',
  NUMBERS_REQUIRED: 'Please type numbers',
};

export const emailPattern = /^[\w-.]+@([\w-]+\.)+[\w-]{2,4}$/g;
export const phonePattern = /\d{1,3}-\d{3}-\d{4}/;

export const initNotification = {
  status: null,
  message: null,
};
