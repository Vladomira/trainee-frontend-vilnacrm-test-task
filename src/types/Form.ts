export type FormDataProps = {
  name: string;
  email: string;
  phone: string;
  address: string;
};
export type FormFieldEvent = { target: { name: string; value: string } };
