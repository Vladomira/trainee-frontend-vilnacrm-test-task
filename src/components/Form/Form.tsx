import Box from '@mui/material/Box';
import Button from '@mui/material/Button';
import Grid from '@mui/material/Grid2';
import TextField from '@mui/material/TextField';
import React, { useState } from 'react';

import { FormFieldEvent, FormDataProps } from '@/types/Form';
import { NotificationInstance } from '@/types/Notification';
import handleApiError from 'pages/api/userService/handleError';
import updateUser from 'pages/api/userService/updateUser';

import Notification from '../Notification';

const emailPattern = /^[\w-.]+@([\w-]+\.)+[\w-]{2,4}$/g;
const phonePattern = /\d{1,3}-\d{3}-\d{4}/;

type FormProps = {
  user: FormDataProps;
};

function Form({ user }: FormProps) {
  const [formData, setFormData] = useState(user);
  const [errors, setErrors] = useState<{ [key: string]: string }>({});
  const [notification, setNotification] = useState<NotificationInstance>({
    status: null,
    message: null,
  });

  const handleChange = (e: FormFieldEvent) => {
    const { name, value } = e.target;
    if (name === 'phone' && /[a-zA-Z]/.test(value)) {
      setErrors((prev) => ({ ...prev, [name]: 'Please type numbers' }));
    } else setFormData((prev) => ({ ...prev, [name]: value }));
  };
  const handleBlur = (e: FormFieldEvent) => {
    const { name, value } = e.target;

    switch (name) {
      case 'email':
        if (!value.match(emailPattern) && value.length > 0) {
          setErrors((prev) => ({ ...prev, [name]: 'Invalid email format' }));
        } else if (value.length === 0) {
          setErrors((prev) => ({ ...prev, [name]: 'Please type email' }));
        }
        break;
      case 'phone':
        if (!value.match(phonePattern) && value.length > 0)
          setErrors((prev) => ({ ...prev, [name]: 'Wrong phone format' }));
        break;

      case 'name':
        if (value.length === 0) setErrors((prev) => ({ ...prev, [name]: 'Please type your name' }));
        break;

      default:
        break;
    }
  };
  const errorReset = (e: FormFieldEvent) => {
    const { name } = e.target;

    setErrors((prev) => {
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      const { [name]: _, ...rest } = prev;
      return rest;
    });
  };
  const formDataReset = (e: FormFieldEvent) => {
    const { name } = e.target;
    setFormData((prev) => ({ ...prev, [name]: '' }));
  };

  const onHandleSubmit = async (e: { preventDefault: () => void }) => {
    e.preventDefault();
    try {
      const result = await updateUser({ id: 1, user: formData });
      const message = `${result.message || 'Update complete'}.User ${
        result.userName || 'Unknown'
      } was updated successfully`;

      setNotification({
        message,
        status: result.status,
      });
    } catch (error: unknown) {
      const { status, message } = handleApiError(error);
      setNotification({ status, message });
    }

    setFormData(user);
    setErrors({});
  };

  return (
    <Box
      component="form"
      padding="50px 0px 0px"
      sx={{ '& .MuiTextField-root': { m: 1, width: '25ch' } }}
    >
      {notification && (
        <Notification setNotification={setNotification} notification={notification} />
      )}
      <Grid
        container
        direction="column"
        display="flex"
        justifyContent="center"
        alignItems="center"
        size="grow"
        spacing={2}
        columns={16}
      >
        <Grid size="auto">
          <TextField
            size="small"
            name="name"
            value={formData.name}
            required
            type="text"
            variant="outlined"
            onChange={handleChange}
            label="Name"
            onBlur={handleBlur}
            error={Boolean(errors.name)}
            helperText={errors.name}
            onFocus={(e) => {
              formDataReset(e);
              errorReset(e);
            }}
          />
        </Grid>
        <Grid size="auto">
          <TextField
            size="small"
            name="email"
            value={formData.email}
            required
            type="email"
            variant="outlined"
            label="Email"
            error={Boolean(errors.email)}
            onBlur={handleBlur}
            onFocus={(e) => {
              formDataReset(e);
              errorReset(e);
            }}
            helperText={errors.email}
            onChange={handleChange}
          />
        </Grid>
        <Grid size="auto">
          <TextField
            size="small"
            name="phone"
            placeholder="380-999-9999"
            value={formData.phone}
            type="tel"
            variant="outlined"
            label="Phone number"
            onChange={handleChange}
            onFocus={(e) => {
              formDataReset(e);
              errorReset(e);
            }}
            error={Boolean(errors.phone)}
            helperText={errors.phone}
            onBlur={handleBlur}
          />
        </Grid>
        <Grid>
          <TextField
            size="medium"
            name="address"
            value={formData.address}
            type="text"
            multiline
            minRows={2}
            maxRows={4}
            placeholder="Street, suite, city"
            variant="outlined"
            onChange={handleChange}
          />
        </Grid>

        <Button
          variant="outlined"
          disabled={!formData || !formData.email || Boolean(errors.email)}
          size="medium"
          type="submit"
          onClick={onHandleSubmit}
        >
          Submit
        </Button>
      </Grid>
    </Box>
  );
}

export default Form;
