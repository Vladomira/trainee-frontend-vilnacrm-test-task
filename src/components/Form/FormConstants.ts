import { FormFieldsData } from '@/types/Form';
import { ErrorInstance } from '@/types/Notification';

export const FORM_ERROR_MESSAGES = {
  INVALID_EMAIL: 'Invalid email format',
  EMAIL_REQUIRED: 'Please type email',
  INVALID_PHONE: 'Wrong phone format',
  NAME_REQUIRED: 'Please type your name',
  INVALID_NAME: 'Only letters are allowed',
  NUMBERS_REQUIRED: 'Please type numbers',
};

export const emailPattern = /^[\w-.]+@([\w-]+\.)+[\w-]{2,4}$/g;
export const phonePattern = /\d{1,3}-\d{3}-\d{4}/;

export const initNotification: ErrorInstance = {
  status: null,
  message: null,
};

export const initialUser: FormFieldsData = {
  name: '',
  email: '',
  phone: '',
  address: '',
};

type FieldsArray = {
  name: keyof FormFieldsData;
  type: string;
  label: string;
  placeholder: string;
}[];
export const fields: FieldsArray = [
  { name: 'name', type: 'text', label: 'Name', placeholder: 'John Doe' },
  { name: 'email', type: 'email', label: 'Email', placeholder: 'example@gmail.com' },
  { name: 'phone', type: 'tel', label: 'Phone number', placeholder: '1-222-333-4444' },
  { name: 'address', type: 'text', label: 'Address', placeholder: 'Address' },
];
