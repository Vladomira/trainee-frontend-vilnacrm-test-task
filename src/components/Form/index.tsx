import React, { useEffect, useState } from 'react';

import { FormDataProps } from '@/types/Form';
import { NotificationInstance } from '@/types/Notification';
import fetchUser from 'pages/api/userService/fetchUser';
import handleApiError from 'pages/api/userService/handleError';

import Notification from '../Notification';

import Form from './Form';

export const initFormData: FormDataProps = {
  name: '',
  email: '',
  phone: '',
  address: '',
};
function FormComponent() {
  const [defaultUser, setDefaultUser] = useState<FormDataProps>(initFormData);
  const [error, setError] = useState<NotificationInstance>({ status: null, message: null });

  useEffect(() => {
    const getUserData = async () => {
      try {
        const result = await fetchUser(1);
        const { name, email, phone, address } = result;

        const { city, street, suite } = address;

        setDefaultUser({
          name,
          email,
          phone: phone.split('x')[0],
          address: `${city}, ${street},${suite}`,
        });
      } catch (err) {
        const { status, message } = handleApiError(error);
        setError({ status, message });
      }
    };

    getUserData();
  }, []);

  return (
    <>
      {error && <Notification notification={error} setNotification={setError} />}
      {defaultUser.name.length > 0 && <Form user={defaultUser} />}
    </>
  );
}

export default FormComponent;
