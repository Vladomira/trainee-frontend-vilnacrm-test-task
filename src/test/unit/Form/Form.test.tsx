// eslint-disable-next-line import/no-extraneous-dependencies
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import React from 'react';

import Form from '@/components/Form/Form';
import SaveButton from '@/components/Form/SaveButton';
import { FORM_ERROR_MESSAGES } from '@/stores/FormConstants';

import { formDataUser } from './store';

describe('Form', () => {
  // email
  it('no error message is shown when email input is valid', async () => {
    render(<Form user={formDataUser} />);

    const emailInput = screen.getByTestId('email');
    const emailError = FORM_ERROR_MESSAGES.INVALID_EMAIL;

    await waitFor(() => {
      expect(emailInput).toBeInTheDocument();
    });

    expect(screen.queryByText(emailError)).not.toBeInTheDocument();
  });
  it('displays an error message when an invalid email is provided', async () => {
    render(<Form user={{ ...formDataUser, email: 'invalid-email' }} />);

    const emailInput = screen.getByDisplayValue('invalid-email');
    const errorMessage = FORM_ERROR_MESSAGES.INVALID_EMAIL;

    await waitFor(() => {
      expect(emailInput).toBeInTheDocument();
    });

    fireEvent.blur(emailInput);
    expect(screen.getByText(errorMessage)).toBeInTheDocument();
  });
  // name
  it('no error message is shown when name is provided', async () => {
    render(<Form user={formDataUser} />);

    const nameInput = screen.getByTestId('name');
    const nameError = FORM_ERROR_MESSAGES.NAME_REQUIRED;

    await waitFor(() => {
      expect(nameInput).toBeInTheDocument();
    });

    expect(screen.queryByText(nameError)).not.toBeInTheDocument();
  });
  // // phone
  it('displays an error message when an invalid phone format is provided', async () => {
    render(<Form user={{ ...formDataUser, phone: '123456789' }} />);

    const phoneInput = screen.getByDisplayValue('123456789');
    const errorMessage = FORM_ERROR_MESSAGES.INVALID_PHONE;

    await waitFor(() => {
      expect(phoneInput).toBeInTheDocument();
    });

    fireEvent.blur(phoneInput);
    expect(screen.getByText(errorMessage)).toBeInTheDocument();
  });
  // submit button
  it('the submit button triggers the correct function', async () => {
    const mockSubmitHandler = jest.fn();

    render(<SaveButton isDisabled={false} onHandleSubmit={mockSubmitHandler} />);
    const saveButton = screen.getByTestId('save-button');

    fireEvent.click(saveButton);

    expect(mockSubmitHandler).toHaveBeenCalledTimes(1);
  });
  it('does not trigger the onClick function when disabled', () => {
    const mockSubmitHandler = jest.fn();
    render(<SaveButton isDisabled onHandleSubmit={mockSubmitHandler} />);

    const button = screen.getByTestId('save-button');

    fireEvent.click(button);

    expect(mockSubmitHandler).not.toHaveBeenCalled();
  });
  it('renders the correct label text', () => {
    render(<SaveButton isDisabled={false} onHandleSubmit={jest.fn()} />);

    expect(screen.getByText('Save')).toBeInTheDocument();
  });
  it('should render fields of form', () => {
    render(<Form user={formDataUser} />);

    expect(screen.getByTestId('name')).toBeInTheDocument();
    expect(screen.getByTestId('email')).toBeInTheDocument();
    expect(screen.getByTestId('phone')).toBeInTheDocument();
    expect(screen.getByTestId('address')).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    const { container } = render(<Form user={formDataUser} />);

    expect(container).toMatchSnapshot();
  });
});
