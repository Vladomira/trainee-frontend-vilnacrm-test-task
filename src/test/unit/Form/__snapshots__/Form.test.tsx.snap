// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should match snapshot 1`] = `
<div>
  <div
    class="formComponentWrapper"
  >
    <form
      class="formContainer"
      data-testid="form"
    >
      <div
        class="MuiGrid2-root MuiGrid2-container MuiGrid2-direction-xs-row MuiGrid2-spacing-xs-3 centredFlexContainer css-12gchgm-MuiGrid2-root"
      >
        <div
          class="MuiGrid2-root MuiGrid2-direction-xs-row MuiGrid2-grid-xs-12 MuiGrid2-grid-md-6 formInput css-1odx5a8-MuiGrid2-root"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-1xp5r68-MuiFormControl-root-MuiTextField-root"
            data-testid="name"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-hvjq6j-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              for=":rk:"
              id=":rk:-label"
            >
              Name
              <span
                aria-hidden="true"
                class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-1ljffdk-MuiFormLabel-asterisk"
              >
                 
                *
              </span>
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                autocomplete="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1pzfmz2-MuiInputBase-input-MuiOutlinedInput-input"
                id=":rk:"
                name="name"
                required=""
                type="text"
                value="Leanne Graham"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-w1u3ce"
                >
                  <span>
                    Name
                     
                    *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid2-root MuiGrid2-direction-xs-row MuiGrid2-grid-xs-12 MuiGrid2-grid-md-6 formInput css-1odx5a8-MuiGrid2-root"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-1xp5r68-MuiFormControl-root-MuiTextField-root"
            data-testid="email"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-hvjq6j-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              for=":rl:"
              id=":rl:-label"
            >
              Email
              <span
                aria-hidden="true"
                class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-1ljffdk-MuiFormLabel-asterisk"
              >
                 
                *
              </span>
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                autocomplete="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1pzfmz2-MuiInputBase-input-MuiOutlinedInput-input"
                id=":rl:"
                name="email"
                required=""
                type="email"
                value="Sincere@april.biz"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-w1u3ce"
                >
                  <span>
                    Email
                     
                    *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid2-root MuiGrid2-direction-xs-row MuiGrid2-grid-xs-12 MuiGrid2-grid-md-6 formInput css-1odx5a8-MuiGrid2-root"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-1xp5r68-MuiFormControl-root-MuiTextField-root"
            data-testid="phone"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-hvjq6j-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              for=":rm:"
              id=":rm:-label"
            >
              Phone number
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                autocomplete="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1pzfmz2-MuiInputBase-input-MuiOutlinedInput-input"
                id=":rm:"
                name="phone"
                placeholder="380-999-9999"
                type="tel"
                value="1-770-736-8031"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-w1u3ce"
                >
                  <span>
                    Phone number
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid2-root MuiGrid2-direction-xs-row MuiGrid2-grid-xs-12 MuiGrid2-grid-md-6 formInput css-1odx5a8-MuiGrid2-root"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-cmpglg-MuiFormControl-root-MuiTextField-root"
            data-testid="address"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-multiline css-szkf5z-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <textarea
                aria-invalid="false"
                autocomplete="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiInputBase-inputSizeSmall css-msbelq-MuiInputBase-input-MuiOutlinedInput-input"
                id=":rn:"
                name="address"
                placeholder="Address"
                style="height: 0px; overflow: hidden;"
              >
                Kulas Light, Apt. 556, Gwenborough
              </textarea>
              <textarea
                aria-hidden="true"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiInputBase-inputSizeSmall css-msbelq-MuiInputBase-input-MuiOutlinedInput-input"
                readonly=""
                style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding-top: 0px; padding-bottom: 0px; width: 100%;"
                tabindex="-1"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-w4cd9x"
                >
                  <span
                    aria-hidden="true"
                    class="notranslate"
                  >
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary saveButton css-gdl1fq-MuiButtonBase-root-MuiButton-root"
          data-testid="save-button"
          tabindex="0"
          type="submit"
        >
          Save
        </button>
      </div>
    </form>
  </div>
</div>
`;
